src/main/java/org/myrobotlab/swing/ProgramABGui.java

```diff
@@ -178,23 +168,74 @@ public void actionPerformed(ActionEvent event) {
 
   @Override
   public void subscribeGui() {
+    subscribe("publishRequest");
+    subscribe("publishText");
+    subscribe("publishOOBText");
   }
 
   @Override
   public void unsubscribeGui() {
+    unsubscribe("publishText");
+    unsubscribe("publishRequest");
+    unsubscribe("publishOOBText");
+  }
+
+  public void onText(String text) {
+    SwingUtilities.invokeLater(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          responseKit.insertHTML(responseDoc, responseDoc.getLength(), "<i><b>" + botName.getText() + "</b>: " + text.replaceAll("\\<.*?\\>", " ").trim() + "</i>", 0, 0, null);
+        } catch (Exception e) {
+          log.error("ProgramAB onText error : {}", e);
+        }
+      }
+    });
+  }
+
+  public void onRequest(String text) {
+    SwingUtilities.invokeLater(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          responseKit.insertHTML(responseDoc, responseDoc.getLength(), "<b>" + userName.getText() + "</b>: " + text, 0, 0, null);
+        } catch (Exception e) {
+          log.error("ProgramAB onRequest error : {}", e);
+        }
+      }
+    });
+  }
+
+  public void onOOBText(String text) {
+    SwingUtilities.invokeLater(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          String filteredOOB = "";
+          Pattern oobPattern = Pattern.compile("<param>(.+?)</param>");

```

Comment:
Use the oobPattern that is already defined in ProgramAB, don't re-define it here.  This pattern should be defined in 1 place and 1 place only.

Reply:
maybe the name can be confusing, Here I don't want to get oobPattern from service ( and we shouldn't ).
I receive String from OOBText publisher, and just format text for needed informations.